Bluetooth: Host: Inconsistent ifdefs for `sc_indicate` can fail to build

_Originally posted by @alxelax in https://github.com/zephyrproject-rtos/zephyr/issues/65216#issuecomment-1812158076_

Declaration and implementation have different
KConfig option dependencies.

Declaration of sc_indicate depends on on set of config  options:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L857-L859
but implementation from a bit different set of config options:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L1518-L1543
then usage of the function happens even there is no body implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L5760-L5767

As a consequence, I get:
```
[15/267] Performing install step for 'tfm'
-- Install configuration: "MinSizeRel"
[260/267] Building C object zephyr/subsys/bluetooth/host/CMakeFiles/subsys__bluetooth__host.dir/gatt.c.obj
/home/alkh-local/ncs/zephyr/subsys/bluetooth/host/gatt.c: In function 'bt_gatt_encrypt_change':
/home/alkh-local/ncs/zephyr/subsys/bluetooth/host/gatt.c:5765:17: warning: implicit declaration of function 'sc_indicate'; did you mean 'gatt_indicate'? [-Wimplicit-function-declaration]
 5765 |                 sc_indicate(0x0001, 0xffff);
      |                 ^~~~~~~~~~~
      |                 gatt_indicate

```
when I build mesh sample with these options:
```
...
CONFIG_SETTINGS=y
CONFIG_BT_SETTINGS=y
...
CONFIG_BT_GATT_CACHING=n
CONFIG_BT_MESH_PB_GATT=n
CONFIG_BT_MESH_GATT_PROXY=n
...

```
for Host it is transformed to:
```
CONFIG_BT_GATT_DYNAMIC_DB=n
CONFIG_BT_GATT_CACHING=n
CONFIG_BT_SETTINGS=y
```
There is declaration and usage of `sc_indicate` but there is no body.
Any suggestions how to fix it on more correct manner are welcome.

--- a/subsys/bluetooth/host/gatt.c
+++ b/subsys/bluetooth/host/gatt.c
@@ -854,9 +854,7 @@ static void db_hash_gen(void)
 	atomic_set_bit(gatt_sc.flags, DB_HASH_VALID);
 }
 
-#if defined(CONFIG_BT_SETTINGS)
 static void sc_indicate(uint16_t start, uint16_t end);
-#endif
 
 static void do_db_hash(void)
 {
@@ -1515,10 +1513,10 @@ void bt_gatt_init(void)
 #endif /* CONFIG_BT_SETTINGS && CONFIG_BT_SMP */
 }
 
-#if defined(CONFIG_BT_GATT_DYNAMIC_DB) || \
-    (defined(CONFIG_BT_GATT_CACHING) && defined(CONFIG_BT_SETTINGS))
 static void sc_indicate(uint16_t start, uint16_t end)
 {
+#if defined(CONFIG_BT_GATT_DYNAMIC_DB) ||                                                          \
+	(defined(CONFIG_BT_GATT_CACHING) && defined(CONFIG_BT_SETTINGS))
 	LOG_DBG("start 0x%04x end 0x%04x", start, end);
 
 	if (!atomic_test_and_set_bit(gatt_sc.flags, SC_RANGE_CHANGED)) {
@@ -1539,8 +1537,8 @@ submit:
 
 	/* Reschedule since the range has changed */
 	sc_work_submit(SC_TIMEOUT);
-}
 #endif /* BT_GATT_DYNAMIC_DB || (BT_GATT_CACHING && BT_SETTINGS) */
+}
 
 void bt_gatt_cb_register(struct bt_gatt_cb *cb)
 {
@@ -5757,14 +5755,12 @@ void bt_gatt_encrypt_change(struct bt_conn *conn)
 
 	bt_gatt_foreach_attr(0x0001, 0xffff, update_ccc, &data);
 
-#if defined(CONFIG_BT_SETTINGS) && defined(CONFIG_BT_GATT_SERVICE_CHANGED)
 	if (!bt_gatt_change_aware(conn, false)) {
 		/* Send a Service Changed indication if the current peer is
 		 * marked as change-unaware.
 		 */
 		sc_indicate(0x0001, 0xffff);
 	}
-#endif	/* CONFIG_BT_SETTINGS && CONFIG_BT_GATT_SERVICE_CHANGED */
 }
 
 bool bt_gatt_change_aware(struct bt_conn *conn, bool req)

